### PR TITLE
Treat `this` as a parameter in IR generation

### PR DESCRIPTION
**Pull request by rdmarsh2**
This stops generating `InitalizeThis` instructions for C++, in favor of treating `this` as a parameter, with corresponding `InitializeParameter` and `InitializeIndirection` instructions. This makes alias analysis handle the `this` pointer accurately. C# IR generation is not currently affected.

This closes https://github.com/github/codeql-c-analysis-team/issues/50

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasConfiguration.qll [hunk 1]
```diff
@@ -5,7 +5,7 @@ private import AliasAnalysis
 
 private newtype TAllocation =
   TVariableAllocation(IRVariable var) or
-  TIndirectParameterAllocation(IRAutomaticUserVariable var) {
+  TIndirectParameterAllocation(IRVariable var) {
```
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasConfiguration.qll:8]:
```suggestion
  TIndirectParameterAllocation(IRAutomaticVariable var) {
```

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasConfiguration.qll [hunk 2]
```diff
@@ -74,7 +74,7 @@ class VariableAllocation extends Allocation, TVariableAllocation {
 }
 
 class IndirectParameterAllocation extends Allocation, TIndirectParameterAllocation {
-  IRAutomaticUserVariable var;
+  IRVariable var;
```
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasConfiguration.qll:77]:
```suggestion
  IRAutomaticVariable var;
```

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/InstructionTag.qll [hunk 3]
```diff
@@ -2,7 +2,10 @@ private import cpp
 
 newtype TInstructionTag =
   OnlyInstructionTag() or // Single instruction (not including implicit Load)
+  InitializeThisAddressTag() or
```
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/InstructionTag.qll:5]: You could probably just reuse the tags we already use when generating other `InitializeParameter` sequences. Not a big deal either way, though.

**Comment by rdmarsh2** [comment]
It does affect one result in dataflow, but this doesn't add the output side. I think it should affect a few more tests, though...

**Comment by MathiasVP** [comment]
> Does this really not affect any dataflow results (yet)?

Apparently the reason we're not getting a lot of dataflow changes is because a setter like this:

```cpp
struct A {
  int x; int y;
  void set_x(int val) {
    this->x = val;
  }
};
```
is not getting a `ReturnIndirection` from which the partial definition of `this` gets propagated out to the caller. With this PR I would expect that setter to behave as the following code, though:

```cpp
void set_x(A* a, int val) {
  a->x = val;
}
```

But the IR for the first setter is:

```
#  118| void A::set_x(int)
#  118|   Block 0
#  118|     v118_1(void)           = EnterFunction                :
#  118|     m118_2(unknown)        = AliasedDefinition            :
#  118|     m118_3(unknown)        = InitializeNonLocal           :
#  118|     m118_4(unknown)        = Chi                          : total:m118_2, partial:m118_3
#  118|     r118_5(glval<unknown>) = VariableAddress[#this]       :
#  118|     m118_6(glval<A>)       = InitializeParameter[#this]   : &:r118_5
#  118|     r118_7(glval<A>)       = Load                         : &:r118_5, m118_6
#  118|     m118_8(A)              = InitializeIndirection[#this] : &:r118_7
#  118|     r118_9(glval<int>)     = VariableAddress[val]         :
#  118|     m118_10(int)           = InitializeParameter[val]     : &:r118_9
#  119|     r119_1(glval<int>)     = VariableAddress[val]         :
#  119|     r119_2(int)            = Load                         : &:r119_1, m118_10
#  119|     r119_3(glval<unknown>) = VariableAddress[#this]       :
#  119|     r119_4(A *)            = Load                         : &:r119_3, m118_6
#  119|     r119_5(glval<int>)     = FieldAddress[x]              : r119_4
#  119|     m119_6(int)            = Store                        : &:r119_5, r119_2
#  119|     m119_7(unknown)        = Chi                          : total:m118_8, partial:m119_6
#  120|     v120_1(void)           = NoOp                         :
#  118|     v118_11(void)          = ReturnVoid                   :
#  118|     v118_12(void)          = AliasedUse                   : m118_3
#  118|     v118_13(void)          = ExitFunction                 :
```

and for the second one it's:

```
#  124| void set_x(A*, int)
#  124|   Block 0
#  124|     v124_1(void)       = EnterFunction            :
#  124|     m124_2(unknown)    = AliasedDefinition        :
#  124|     m124_3(unknown)    = InitializeNonLocal       :
#  124|     m124_4(unknown)    = Chi                      : total:m124_2, partial:m124_3
#  124|     r124_5(glval<A *>) = VariableAddress[a]       :
#  124|     m124_6(A *)        = InitializeParameter[a]   : &:r124_5
#  124|     r124_7(A *)        = Load                     : &:r124_5, m124_6
#  124|     m124_8(unknown)    = InitializeIndirection[a] : &:r124_7
#  124|     r124_9(glval<int>) = VariableAddress[val]     :
#  124|     m124_10(int)       = InitializeParameter[val] : &:r124_9
#  125|     r125_1(glval<int>) = VariableAddress[val]     :
#  125|     r125_2(int)        = Load                     : &:r125_1, m124_10
#  125|     r125_3(glval<A *>) = VariableAddress[a]       :
#  125|     r125_4(A *)        = Load                     : &:r125_3, m124_6
#  125|     r125_5(glval<int>) = FieldAddress[x]          : r125_4
#  125|     m125_6(int)        = Store                    : &:r125_5, r125_2
#  125|     m125_7(unknown)    = Chi                      : total:m124_8, partial:m125_6
#  126|     v126_1(void)       = NoOp                     :
#  124|     v124_11(void)      = ReturnIndirection[a]     : &:r124_7, m125_7
#  124|     v124_12(void)      = ReturnVoid               :
#  124|     v124_13(void)      = AliasedUse               : m124_3
#  124|     v124_14(void)      = ExitFunction             :
```

Is that intentional?

**Comment by rdmarsh2** [comment]
Yes. I was planning to add the `ReturnIndirection` in a followup PR.

**Comment by rdmarsh2** [comment]
It looks like the same issue as the existing two failures (Chi node associated with a call argument that's a pointer to a field having both the pointer type and an unknown type)

**Pull request merged**